rule get_files_for_multiqc_paired:
    input:
        ancient("strains/{sample}/quality/qualimap/qualimapReport.html"),
        ancient("strains/{sample}/quality/quast/report.txt"),
        ancient("strains/{sample}/reads/trimmed/fastqc/R1_paired_fastqc/summary.txt"),
        ancient("strains/{sample}/annotation/{sample}.txt"),
        ancient("strains/{sample}/logs/trimmomatic.log"),
#        "strains/{sample}/contamination/blast_results_taxid_species.tsv"
    output:
        "strains/{sample}/multiqc/log.txt"
    shell:
        """
        mkdir -p $(dirname {output[0]})/{wildcards.sample}
        cp -R $(dirname {input[0]})/* $(dirname {output[0]})/{wildcards.sample}
        sed "s/_R1\.fastq/\.fastq/g" {input[4]} | sed "s/_R2\.fastq/\.fastq/g" > $(dirname {output[0]})/trimmomatic.log
        cp {input[3]} $(dirname {output[0]})
        cp -R $(dirname {input[2]})/* $(dirname {output[0]})/{wildcards.sample}
        sed -i.bak "s/R1_paired\.fastq/{wildcards.sample}/" $(dirname {output[0]})/{wildcards.sample}/fastqc_data.txt
        sed -i.bak "s/R2_paired\.fastq/{wildcards.sample}/" $(dirname {output[0]})/{wildcards.sample}/fastqc_data.txt
        sed "s/{wildcards.sample}\.fsa/{wildcards.sample}/g" {input[1]} > $(dirname {output[0]})/quast.log
        cp -R $(dirname {input[1]}) $(dirname {output[0]})
        sed -i.bak "s/.fsa//g" $(dirname {output[0]})/quast/report.tsv
        echo "okay" > {output}
        """


rule get_files_for_multiqc_single:
    input:
        "strains/{sample}/quality/qualimap/qualimapReport.html",
        "strains/{sample}/quality/quast/report.txt",
        "strains/{sample}/reads/trimmed/fastqc/single_fastqc/summary.txt",
        "strains/{sample}/annotation/{sample}.txt",
        "strains/{sample}/logs/trimmomatic.log",
#        "strains/{sample}/contamination/blast_results_taxid_species.tsv"

    output:
        "strains/{sample}/multiqc/log.txt"
    shell:
        """
        mkdir -p $(dirname {output[0]})/{wildcards.sample}
        cp -R $(dirname {input[0]})/* $(dirname {output[0]})/{wildcards.sample}
        sed "s/_R1\.fastq/\.fastq/g" {input[4]} | sed "s/_R2\.fastq/\.fastq/g" > $(dirname {output[0]})/trimmomatic.log
        cp {input[3]} $(dirname {output[0]})
        cp -R $(dirname {input[2]})/* $(dirname {output[0]})/{wildcards.sample}
        sed -i.bak "s/R1_paired\.fastq/{wildcards.sample}/" $(dirname {output[0]})/{wildcards.sample}/fastqc_data.txt
        sed -i.bak "s/R2_paired\.fastq/{wildcards.sample}/" $(dirname {output[0]})/{wildcards.sample}/fastqc_data.txt
        sed "s/{wildcards.sample}\.fsa/{wildcards.sample}/g" {input[1]} > $(dirname {output[0]})/quast.log
        cp -R $(dirname {input[1]}) $(dirname {output[0]})
        sed -i.bak "s/.fsa//g" $(dirname {output[0]})/quast/report.tsv
        echo "okay" > {output}
        """


        

rule multiqc_all_to_self_genome:
    conda:
        "env/multiqc.yaml"
    input:
        expand("strains/{sample}/multiqc/log.txt", sample = read_naming.keys())
    output:
        "quality/multiqc_report.html"
    shell:
        """
        multiqc -f -o $(dirname {output[0]}) strains/*/multiqc/
        """




rule get_files_for_multiqc_to_reference:
    input:
        "strains/{sample}/quality/mapping/bwa/{ref}/qualimap/qualimapReport.html"
    output:
        "strains/{sample}/multiqc/mapping_to_{ref}/log.txt"
    shell:
        """
        mkdir -p $(dirname {output[0]})/{wildcards.sample}
        cp -R $(dirname {input[0]})/* $(dirname {output[0]})/{wildcards.sample}
        echo "okay" > {output}
        """


rule multiqc_all_to_reference:
    conda:
        "env/multiqc.yaml"
    input:
        expand("strains/{sample}/multiqc/mapping_to_{{ref}}/log.txt", sample = read_naming.keys())
    output:
        "quality/multiqc_mapping_to_{ref}/multiqc_report.html"
    shell:
        """
        multiqc -f -o $(dirname {output[0]}) strains/*/multiqc/mapping_to_{wildcards.ref}/
        """

