rule qualimap_mapping:
    conda:
        "env/qualimap.yaml"
    input:
        "strains/{sample}/mapping/qualimap/{sample}.bam",
        "strains/{sample}/annotation/{sample}_truncated.gff"
    output:
        "strains/{sample}/mapping/qualimap/qualimapReport.html"
    shell:
        """
        qualimap bamqc -bam {input[0]} -gff {input[1]} -outdir $(dirname {output})
        """

rule quast:
    conda:
        "env/quast.yaml"
    input:
        "strains/{sample}/annotation/{sample}.fsa"
    output:
        "strains/{sample}/assembly/spades/quast/report.txt"
    shell:
        "quast.py {input} -o $( dirname {output})"

        
rule convert_quast_images:
    conda:
        "env/imagemagick.yaml"
    input:
        "strains/{sample}/assembly/spades/quast/report.txt"
    output:
        "strains/{sample}/assembly/spades/quast/basic_stats/contigs_500bp_high_kmer_coverage_GC_content_plot.png",
        "strains/{sample}/assembly/spades/quast/basic_stats/contigs_500bp_high_kmer_coverage_coverage_histogram.png"
    shell:
        """
        convert $(dirname {input[0]})/basic_stats/contigs_500bp_high_kmer_coverage_coverage_histogram.pdf $(dirname {input[0]})/basic_stats/contigs_500bp_high_kmer_coverage_coverage_histogram.png
        convert $(dirname {input[0]})/basic_stats/contigs_500bp_high_kmer_coverage_GC_content_plot.pdf $(dirname {input[0]})/basic_stats/contigs_500bp_high_kmer_coverage_GC_content_plot.png
        """


rule check_coverage_assembly:
    input:
        "strains/{sample}/assembly/spades/contigs_500bp.fasta"
    output:
        "strains/{sample}/assembly/spades/contigs_500bp_low_kmer_coverage.txt",
        "strains/{sample}/assembly/spades/contigs_500bp_high_kmer_coverage.txt"
    shell:
        "grep \">\" {input} | sed \"s/.*cov_//\" | awk '$1 < 5 {{print NR}}' | sed \"s/^/NODE_/\" | sed \"s/$//\" | sed \"s/^>//\" > {output[0]} && grep \">\" {input} | sed \"s/.*cov_//\" | awk '$1 > 5 {{print NR}}' | sed \"s/^/NODE_/\" | sed \"s/$/_/\" | sed \"s/^>//\"  > {output[1]} "

        
rule filter_contigs_on_coverage:
    input:
        "strains/{sample}/assembly/spades/contigs_500bp.fasta",
        "strains/{sample}/assembly/spades/contigs_500bp_high_kmer_coverage.txt",
        "strains/{sample}/assembly/spades/contigs_500bp_low_kmer_coverage.txt"
    output:
        "strains/{sample}/assembly/spades/contigs_500bp_high_kmer_coverage.fasta",
        "strains/{sample}/assembly/spades/contigs_500bp_low_kmer_coverage.fasta"
    shell:
        """
        grep -A 1 -f {input[1]} {input[0]} | sed '/^--$/d' > {output[0]}
        if [ -s {input[2]} ]
        then 
            grep -A 1 -f {input[2]} {input[0]} | sed '/^--$/d' > {output[1]}
        else
            touch {output[1]}
        fi
        """
