rule download_reference_from_url:
    params:
        id = species_taxid,
    conda:
        "../../envs/entrez-direct.yaml"
    output:
        genome="references/{ref}/genome_{format}.{format}",
        subvalue="references/{ref}/genome_subvalue.txt",
    shell:
        """
        ftp=$(efetch -db assembly -id {wildcards.ref} -format docsum | xtract -pattern DocumentSummary -unless LatestAccession -if AssemblyStatus -equals "Complete Genome"  -if SpeciesTaxid -equals {params.id} -if FtpPath_RefSeq -element FtpPath_RefSeq)
        nucl_id=$(echo ${{ftp}} | sed  "s/\(\/GCF_.*\)/\\1\\1_assembly_report.txt/" | xargs wget -qO- | sed '/^#.*/ d'  | awk ' $4 == "Chromosome" {{print $7}}')
        efetch -db nucleotide -id ${{nucl_id}} -format {wildcards.format} > {output[genome]} 
        efetch -db assembly -id {wildcards.ref} -format docsum | xtract -pattern DocumentSummary -unless LatestAccession -if AssemblyStatus -equals "Complete Genome" -if SpeciesTaxid -equals {params.id} -if FtpPath_RefSeq -element Sub_value > {output[subvalue]}
        """

