rule mash_sketch_complete_genomes:
    conda:
        "../quality/env/mash.yaml"
    input:
        "references/all_complete_genomes_log.txt"
    output:
        "references/all_complete_genomes.msh"
    shell:
        """
        mash sketch -o {output[0]} $(dirname {input[0]})/all_complete_genomes/*.fasta
        """
    
rule mash_dist_sample:
    conda:
        "../quality/env/mash.yaml"
    input:
        "references/all_complete_genomes.msh",
        "strains/{sample}/annotation/{sample}.fsa"
    output:
        "strains/{sample}/closest_genome.txt"
    shell:
        """
        mash dist {input} | sort -gk3 | sed -n 1p | cut -f1 | sed "s/references\/all_complete_genomes\///" | sed "s/_genome\.fasta//" > {output[0]}
        """

rule get_all_closest_genomes:
    input:
        expand("strains/{sample}/closest_genome.txt", sample=read_naming.keys())
    output:
        "closest_genomes.txt"
    shell:
        """
        head {input} | sed "/^$/ d" | awk '{{ORS = (NR%2 ? FS : RS)}} 1' | sed "s/==> strains\///" | sed "s/\/closest_genome.txt <==//" | sort -k2   > {output}
        """

        

rule get_fastas_closest_genomes:
    input:
        "closest_genomes.txt",
        "references/url_complete_genomes.txt"
    output:
        dynamic("reference/closest/{ref}/genome.fasta")
    shell:
        """
        while read strain id
        do
            nucl_id=$(grep "\\b${{id}}\\b" {input[1]} | cut -f2 |  sed  "s/\(\/GCF_.*\)/\\1\\1_assembly_report.txt/" | xargs wget -qO- | sed '/^#.*/ d'  | awk ' $4 == "Chromosome" {{print $7}}')
            echo ${{nucl_id}}
            efetch -db nucleotide -id ${{nucl_id}} -format fasta > {output[0]} 
        done < {input[0]}
        """
