rule remove_fasta_part_from_gff:
    input:
        "strains/{sample}/annotation/{sample}.gff"
    output:
        "strains/{sample}/annotation/{sample}_truncated.gff"
    shell:
        "sed '/^##FASTA$/,$d' {input} > {output}"
    
rule trimmomatic_SE:
    conda:
        "env/trim.yaml"
    input:
        "reads/raw/{sample}.fastq"
    params:
        minlength = config["minimum_read_length"],
        minqual = config["minimum_quality_base"],
        croplength = config["crop_at_read_beginning"]
    output:
        temp("strains/{sample}/reads/trimmed/single.fastq"),
    log:
        "strains/{sample}/logs/trimmomatic.log"
    shell:
        """
        trimmomatic SE {input[0]} {output[0]} ILLUMINACLIP:${{CONDA_DEFAULT_ENV}}/share/trimmomatic/adapters/TruSeq3-PE.fa:3:25:6 LEADING:{params.minqual} TRAILING:{params.minqual} MINLEN:{params.minlength} HEADCROP:{params.croplength} &> {log}
        """

rule trimmomatic_PE:
    conda:
        "env/trim.yaml"
    input:
        "reads/raw/{sample}_R1.fastq",
        "reads/raw/{sample}_R2.fastq"
    params:
        minlength = config["minimum_read_length"],
        minqual = config["minimum_quality_base"],
        croplength = config["crop_at_read_beginning"]
    output:
        temp("strains/{sample}/reads/trimmed/R1_paired.fastq"),
        temp("strains/{sample}/reads/trimmed/R1_unpaired.fastq"),
        temp("strains/{sample}/reads/trimmed/R2_paired.fastq"),
        temp("strains/{sample}/reads/trimmed/R2_unpaired.fastq"),
        "strains/{sample}/logs/trimmomatic.log"
    log:
        "strains/{sample}/logs/trimmomatic.log"
    shell:
        """
        trimmomatic PE {input[0]} {input[1]} {output[0]} {output[1]} {output[2]} {output[3]} ILLUMINACLIP:${{CONDA_DEFAULT_ENV}}/share/trimmomatic/adapters/TruSeq3-PE.fa:3:25:6 LEADING:{params.minqual} TRAILING:{params.minqual} MINLEN:{params.minlength} HEADCROP:{params.croplength} &> {log}
        """        


rule error_correction_spades_paired:
    conda:
        "env/spades.yaml"
    input:
        "strains/{sample}/reads/trimmed/R1_paired.fastq",
        "strains/{sample}/reads/trimmed/R2_paired.fastq"
    output:
        temp("strains/{sample}/reads/corrected/R1_paired.00.0_0.cor.fastq.gz"),
        temp("strains/{sample}/reads/corrected/R2_paired.00.0_0.cor.fastq.gz"),
        temp("strains/{sample}/reads/corrected/R_unpaired.00.0_0.cor.fastq.gz")
    log:
        "strains/{sample}/logs/spades_read_correction.txt"
    shell:
         """
         spades.py -1 {input[0]} -2 {input[1]} --only-error-correction -o $( dirname $(dirname {output[0]})) > {log} 
         """

rule error_correction_spades_single:
    conda:
        "env/spades.yaml"
    input:
        "strains/{sample}/reads/trimmed/single.fastq",
    output:
        temp("strains/{sample}/reads/corrected/single.00.0_0.cor.fastq.gz")
    log:
        "strains/{sample}/logs/spades_read_correction.txt"
    shell:
         "spades.py -s {input[0]} --only-error-correction -o $( dirname $(dirname {output[0]})) > {log} "

         
rule spades_paired:
    conda:
        "env/spades.yaml"
    input:
        ancient("strains/{sample}/reads/corrected/R1_paired.00.0_0.cor.fastq.gz"),
        ancient("strains/{sample}/reads/corrected/R2_paired.00.0_0.cor.fastq.gz"),
        ancient("strains/{sample}/reads/corrected/R_unpaired.00.0_0.cor.fastq.gz")
    output:
        "strains/{sample}/assembly/spades/contigs.fasta"
    log:
        "strains/{sample}/logs/spades_assembly.txt"
    shell:
        """
        spades.py --only-assembler -1 {input[0]} -2 {input[1]} -s {input[2]} -o $( dirname {output[0]}) > {log}
        mv $(dirname {output[0]})/contigs.fasta $(dirname $(dirname {output[0]}))
        rm -rf $( dirname {output[0]})/*
        mv $(dirname $(dirname {output[0]}))/contigs.fasta $(dirname {output[0]})/
        """

rule spades_single:
    conda:
        "env/spades.yaml"
    input:
        "strains/{sample}/reads/corrected/single.00.0_0.cor.fastq.gz",
    output:
        "strains/{sample}/assembly/spades/contigs.fasta"
    log:
        "strains/{sample}/logs/spades_assembly.txt",
    shell:
        """
        spades.py --only-assembler -s {input[0]} -o $( dirname {output[0]}) > {log}
        mv $(dirname {output[0]})/contigs.fasta $(dirname $(dirname {output[0]}))
        rm -rf $( dirname {output[0]})/*
        mv $(dirname $(dirname {output[0]}))/contigs.fasta $(dirname {output[0]})/
        """
        
rule extract_contigs_500bp:
    input:
        "strains/{sample}/assembly/spades/contigs.fasta"
    output:
        "strains/{sample}/assembly/spades/contigs_500bp.fasta"
    shell:
        "awk '/^>/{{print (NR==1)?$0: \"\\n\" $0;next}} {{printf \"%s\", $0}}END{{print \"\"}}' {input} |  awk \'!/^>/ {{ next }} {{ getline seq }} length(seq) >= 500 {{ print $0 \"\\n\" seq }}\'  > {output}"  


rule rename_contigs:
    input:
        "strains/{sample}/assembly/spades/contigs_500bp_high_kmer_coverage.fasta"
    output:
        "strains/{sample}/assembly/spades/contigs_500bp_high_kmer_coverage_renamed.fasta"
    shell:
        "sed \"s/NODE_\\([0-9]\\+\\)_.*/contig00\\1/\" {input} > {output}"

