import csv
import os
import shutil

snakemake_path = "/home/sacha/Documents/new_strain/"

sys.path.insert(0, snakemake_path)

from helper_scripts.get_sample_naming_from_folder_general import get_read_naming_patterns

from helper_scripts.get_folder_and_subfolders import get_folders_structure_with_sample_naming

def get_sample_names_correspondance(filename):
    corres = {}
    with open(filename) as csvfile:
        reader = csv.reader(csvfile, delimiter='\t', quotechar='|')
        next(reader, None)
        next(reader, None)
        for row in [[s.strip() for s in inner] for inner in reader]:
            cells = list(filter(None, row))
            corres[cells[0]] = cells[1].replace(" ", "_")
    return(corres)

link_directory="links/"


read_naming, folders, original_name  = get_folders_structure_with_sample_naming(link_directory, get_read_naming_patterns)


if os.path.isfile("sample_names.tsv"):
    sample_correspondance = get_sample_names_correspondance("sample_names.tsv")
else:
    sample_correspondance = { x:x for x in read_naming.keys()} 

taxid=1773
db_to_use = "myco"
spec_for_mykrobe = "tb"
mysql_conf_file = "/home/sacha/.my.cnf"
sample_names = "sample_names.tsv"
resistance_prediction_softwares = ["mykrobe", "rgi"]
currated_genes_file = "resistance_genes.csv"
species = "Mycobacterium tuberculosis"

genes = []
with open(currated_genes_file) as csvfile:
    reader = csv.reader(csvfile, delimiter="\t")
    for row in reader:
        genes.append(row[1])

genes.append("gidB")

        
ontology_file_aro = 'aro.obo'
ontology_file_mo = 'mo.obo'
ontology_file_ro = 'ro.obo'

prefix = "../../rules/"

include:
    prefix + "read_manipulation/get_reads.rules"

include:
    prefix + "downloading/fetch_references.rules"
    
include:
    prefix + "downloading/fetch_software.rules"

include:
    prefix + "mapping/snippy.rules"

include:
    prefix + "mapping/align_reads_to_genome.rules"

include:
    prefix + "phylogeny/raxml.rules"

include:
    prefix + "assembly/assembly.rules"

include:
    prefix + "db_management/creation.rules"

include:
    prefix + "db_management/insert_data.rules"

include:
    prefix + "db_management/fetch_data.rules"

include:
    prefix + "annotation/prokka.rules"

include:
    prefix + "annotation/resistance.rules"
    
include:
    prefix + "quality/multiqc.rules"

include:
    prefix + "quality/quality_assembly.rules"

include:
    prefix + "quality/quality_reads.rules"    
    
include:
    prefix + "typing/mlst.rules"

include:
    prefix + "typing/snps_distance.rules"
    
include:
    prefix + "typing/compare_methods.rules"

include:
    prefix + "core_genome_calculations/parsnp.rules"

include:
    prefix + "core_genome_calculations/extract_snps_core.rules"



rule something:
    input:
        expand("strains/{sample}/mapping/qualimap/{sample}.bam", sample= read_naming.keys())
#        expand("mapping/snippy/cdc1551/{sample}/snps.vcf", sample= read_naming.keys())
     #
    output:
        "report.html"
    shell:
        """
        echo "" > {output[0]}
        """
    
